### PR TITLE
Puppi 75x miniaodpass2

### PR DESCRIPTION
**Pull request by nhanvtran**
PUPPI update with tuning (neutral pT cuts and slopes) and bug fixes (uniformized charged particle treatment between AOD/miniAOD based on fromPV definition)

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc [hunk 1]
```diff
@@ -70,6 +71,13 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
   iEvent.getByToken(tokenVertices_,hVertexProduct);
   const reco::VertexCollection *pvCol = hVertexProduct.product();
 
+   int npv = 0;
+   const reco::VertexCollection::const_iterator vtxEnd = pvCol->end();
+   for (reco::VertexCollection::const_iterator vtxIter = pvCol->begin(); vtxEnd != vtxIter; ++vtxIter) {
+      if (!vtxIter->isFake() && vtxIter->ndof()>=4 && fabs(vtxIter->z())<=24)
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: magic numbers are not so nice

File: CommonTools/PileupAlgos/src/PuppiContainer.cc [hunk 2]
```diff
@@ -235,14 +237,17 @@ std::vector<double> const & PuppiContainer::puppiWeights() {
...
+        if(fInvert) pWeight = 1.-pWeight;
         //std::cout << "fRecoParticles[i0].pt = " <<  fRecoParticles[i0].pt << ", fRecoParticles[i0].charge = " << fRecoParticles[i0].charge << ", fRecoParticles[i0].id = " << fRecoParticles[i0].id << ", weight = " << pWeight << std::endl;
 
         fWeights .push_back(pWeight);
         fAlphaMed.push_back(fPuppiAlgo[pPupId].median(0));
         fAlphaRMS.push_back(fPuppiAlgo[pPupId].rms(0));        
         //Now get rid of the thrown out weights for the particle collection
-        if(std::abs(pWeight) < std::numeric_limits<double>::denorm_min() ) continue;
+
+        // if( std::abs(pWeight) < std::numeric_limits<double>::denorm_min() ) continue; // this line seems not to work like it's supposed to...
```
Comment (slava77) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc]: is this commented out code still needed?
please cleanup

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc [hunk 3]
```diff
@@ -100,40 +111,72 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
-        pD0        = lPack->dxy();
-        closestVtx = &(*(lPack->vertexRef()));
-        pVtxId = (lPack->fromPV() != (pat::PackedCandidate::PVUsedInFit));
-        if( (lPack->fromPV() == pat::PackedCandidate::PVLoose) || (lPack->fromPV() == pat::PackedCandidate::PVTight) ){
-            pVtxId = 0;
-        } 
-    }
+      int tmpFromPV = 0;  
+      // mocking the miniAOD definitions
+      if (closestVtx != 0 && fabs(pReco.charge) > 0 && pVtxId > 0) tmpFromPV = 0;
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: these if blocks appear pretty wasteful and are probably less readable than nested ones.
Especially when the same condition repeats in every single one.

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc [hunk 4]
```diff
@@ -100,40 +111,72 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
+      if (closestVtx != 0 && fabs(pReco.charge) > 0 && pVtxId == 0) tmpFromPV = 3;
+      if (closestVtx == 0 && fabs(pReco.charge) > 0 && closestVtxForUnassociateds == 0) tmpFromPV = 2;
+      if (closestVtx == 0 && fabs(pReco.charge) > 0 && closestVtxForUnassociateds != 0) tmpFromPV = 1;
       pReco.dZ      = pDZ;
       pReco.d0      = pD0;
+      pReco.id = 0; 
+      if (fabs(pReco.charge) == 0){ pReco.id = 0; }
+      if (fabs(pReco.charge) > 0){
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: why not "else" here instead of ```if (fabs(pReco.charge) > 0) ``` ?

**Comment by slava77** [comment]
Looking in 750 (a. 70 ttbar events, b. 200 Zmm events each case with PU=35@25ns:
- puppi takes about 10% more CPU in each case
- the number of slimmedJetsPuppi reduces substantially. This reduction seems reasonable though.
    - from 10.3 to 5.7 in ttbar
    - from 6.0 to 2.3 in Zmm events
    - Do these numbers make sense?

**Comment by slava77** [comment]
@nhanvtran please confirm the behavior of the slimmedJetsPuppi multiplicity is expected

**Comment by nhanvtran** [comment]
@slava77 The numbers you are quoting, is this the reduction from slimmedJets -> slimmedJetsPuppi?  These are jets > 10 GeV in both cases?  If so, this makes sense, especially since the workflow is not including the right JECs right now (which is a small effect).

Going through the rest of the line comments soon...

**Comment by slava77** [comment]
On 7/30/15 8:29 AM, nhanvtran wrote:
> @slava77 <https://github.com/slava77> The numbers you are quoting, is
> this the reduction from slimmedJets -> slimmedJetsPuppi? 

No, this is a comparison in the number of jets in the slimmedJetsPuppi
collection before and after this PR changes.

**Comment by nhanvtran** [comment]
Ok, sorry for being dense.  Still jets above 10 GeV?  Theforward tune got much better (more aggressive) so this makes sense.

**Comment by slava77** [comment]
Looking at the config, it's after a 20 GeV cut
```
selectedPatJetsPuppi.cut = "pt > 20" 
slimmedJetsPuppi.src = "selectedPatJetsPuppi"
```

I suppose in this case it (the reduction in the number of jets per event) makes even more sense.

**Comment by nhanvtran** [comment]
Ok, yes, I agree.

Switching topics...
Given the discussion at today's miniAOD meeting, I suppose we should keep updating this PR.  Is it okay?

**Comment by slava77** [comment]
I thought it was perfect as it is already.
Isn't it?

**Comment by nhanvtran** [comment]
so modulo your suggestions above, this is nearly complete.  

However, in the meeting today, the plan was to add both Puppi interpretations to the miniAOD (as two weights to each packed candidate) so that Puppi isolation can be run off the miniAOD as well.

**Comment by slava77** [comment]
Thanks for the clarification.
I'm fine with updates of this (and the 74X) PR.

**Comment by slava77** [comment]
@nhanvtran should I expect an update of this PR soon?

**Comment by nhanvtran** [comment]
@slava77 should have something by today, coordinating with @rappoccio

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc (hunk 3 above)
Comment (nhanvtran) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: We've had so many issues syncing miniAOD and AOD that we prefer to keep this code.

File: CommonTools/PileupAlgos/src/PuppiContainer.cc (hunk 2 above)
Comment (nhanvtran) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc]: Would like to keep these 2 lines nin case we break the 1-to-1 ness in the future

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc (hunk 1 above)
Comment (nhanvtran) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: Thanks, made configurable

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc (hunk 4 above)
Comment (nhanvtran) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: done

**Comment by slava77** [comment]
ping

**Comment by slava77** [comment]
ping
this one should be updated (in agreement with #10551 )

**Comment by nhanvtran** [comment]
hi slava, yes, sorry porting the code over today
LPC had some downtime some I'm trying to transfer some things over...

**Pull request closed without merging**